src/sqlfluff/core/templaters/python.py

```diff
@@ -305,6 +305,11 @@ def slice_file(
             else:
                 # If it's not equal, loop around
                 templated_str = new_templated_str
+            if trailing_newline:
+                # Used by dbt.py to determine if trailing newline needs to be
```

Comment:
1. As this is dbt-specific, it should go in `templaters/dbt.py`, before calling slice_file.

2. We should ensure that rather than blindly adding an end of file newline, we add the same number as were in the raw SQL.

Reply:
Same remark on 2., we should have tests checking that:
1. If a model does not contain a new line originally, then the rule fails
2. If a model contains a new line originally, then the rule passes

Reply:
Sorry for the late response to this review @NiallRees and @dmateusp.

@NiallRees 
1. I think this is a case of my comment just causing confusion instead of helping (I plan to remove it).<br>`if trailing_newline:` has to be here since I am simply looking that the argument has been set to `True` when `slice_file()` is called in `dbt.py`(by default I have set it to `False`). This ensures `slice_file()` only considers any newline logic when using the dbt templater and disregards when using the Jinja templater (or wherever else it may be used). You can see how this plays out in `dbt.py` [here](https://github.com/GClunies/sqlfluff/blob/ad559848ce0591d02d713b43bc854c7cb4531367/src/sqlfluff/core/templaters/dbt.py#L255-L262).

2. Completely agree - we should add the same number as were in the raw SQL.

@dmateusp 
1. Agree
2. Agree

Thanks again for the review. I will try and make changes this week.

Reply:
Hey @GClunies no worries, let me try again! I can see that you've added the default argument of False to trailing_newline, but as the only templater that would use the argument is the dbt templater, it makes sense for this logic to go inside the dbt templater. It doesn't need to be in slice_file from what I can see. The dbt templater can just add any new lines it needs to before handing over the raw and templated sql to slice_file. Make sense?

Reply:
Got it, will try this.

Reply:
@NiallRees 

I remember why I ended up handling this issue in `python.py` and not `dbt.py`

If I change [these lines](https://github.com/sqlfluff/sqlfluff/blob/72d3aafc754bc1ecde4682a6624e16d8a97f6c01/src/sqlfluff/core/templaters/dbt.py#L245-L264) in `dbt.py` to
```python
...
...
...
if not compiled_sql:
            raise SQLTemplaterError(
                "dbt templater compilation failed silently, check your configuration "
                "by running `dbt compile` directly."
            )

        raw_sliced, sliced_file, templated_sql = self.slice_file(
            node.raw_sql, compiled_sql, config=config
        )

       templated_sql = templated_sql + "\n"  # ADDING NEWLINE IN dbt.py AS SUGGESTED

        return (
            TemplatedFile(
                source_str=node.raw_sql,
                templated_str=templated_sql,
                fname=fname,
                sliced_file=sliced_file,
                raw_sliced=raw_sliced,
            ),
            # No violations returned in this way.
            [],
        )
```

I get the following error when I run `sqlfluff fix`
```bash
▶ sqlfluff fix models/analytics/editorial/sl_web_editorial_time_on_page.sql
==== finding fixable violations ====
WARNING    Unable to lint models/analytics/editorial/sl_web_editorial_time_on_page.sql due to an internal error. Please report this as an issue with your query's contents and stacktrace below!
To hide this warning, add the failing file to .sqlfluffignore
Traceback (most recent call last):
  File "/Users/gregclunies/Repos/sqlfluff/src/sqlfluff/core/linter.py", line 1271, in lint_path
    target_file.read(), fname=fname, fix=fix, config=config
  File "/Users/gregclunies/Repos/sqlfluff/src/sqlfluff/core/linter.py", line 1060, in lint_string
    parsed = self.parse_string(in_str=in_str, fname=fname, config=config)
  File "/Users/gregclunies/Repos/sqlfluff/src/sqlfluff/core/linter.py", line 810, in parse_string
    tokens, lex_vs = lexer.lex(templated_file)
  File "/Users/gregclunies/Repos/sqlfluff/src/sqlfluff/core/parser/lexer.py", line 349, in lex
    return self.enrich_segments(segment_buff, raw), violations
  File "/Users/gregclunies/Repos/sqlfluff/src/sqlfluff/core/parser/lexer.py", line 379, in enrich_segments
    templated_slice
  File "/Users/gregclunies/Repos/sqlfluff/src/sqlfluff/core/templaters/base.py", line 257, in templated_slice_to_source_slice
    template_slice.stop, inclusive=False
  File "/Users/gregclunies/Repos/sqlfluff/src/sqlfluff/core/templaters/base.py", line 184, in _find_slice_indices_of_templated_pos
    raise ValueError("Position Not Found")
ValueError: Position Not Found

==== no fixable linting violations found ====
```

I think this is because the modified templated SQL resulting from `templated_sql = templated_sql + "\n"` no longer "matches up with" `sliced_file` that is returned from...
```python 
raw_sliced, sliced_file, templated_sql = self.slice_file(
            node.raw_sql, compiled_sql, config=config
        )
```
...which is run just above it and is used by the `TemplatedFile()` class below it.

So seems to me the options are:
1. Solve it as I have in `python.py`. Not ideal in terms of "structure" as you called out, but gets the job done and seems to be the DRY solution.
2. Have an analogous `slice_file()` class method in `dbt.py` that overrides the definition in `python.py`. It would essentially be a carbon copy of what I did in `python.py`, just moved to `dbt.py`. I'll call this the WET solution.💦 
3. A more elegant solution I have not thought of! 😆

Reply:
@NiallRees @dmateusp 

I have updated (and simplified) the code in this PR so that:
1. All of the logic for handling trailing newlines is in `templaters/dbt.py`.
2. Rather than blindly adding a _single_ end of file newline, we add the same number as are in the raw SQL.

I have also updated the PR description above to reflect these changes. TL;DR - users do not have to include exclude_rules = L009 when using the dbt-templater anymore.

I'll be working on tests now that I have a solution I am satisfied with.